lib/FFI/Platypus/Type/Win32/LPCWSTR.pm

```diff
@@ -0,0 +1,132 @@
+package FFI::Platypus::Type::Win32::LPCWSTR;
+
+use strict;
+use warnings;
+use 5.008004;
+use FFI::Platypus;
+use FFI::Platypus::Memory qw( strlenW );
+use Encode qw( decode encode );
+
+# ABSTRACT: Platypus custom type for Windows Unicode strings
+# VERSION
+
+=head1 SYNOPSIS
+
+In your C code:
+
+ void
+ takes_unicode_string(LPCWSTR s)
+ {
+   ...
+ }
+
+In your L<Platypus::FFI> code:
+
+ use FFI::Platypus;
+
+ my $ffi = FFI::Platypus->new( api => 1 );
+ $ffi->load_custom_type('::Win32::LPCWSTR' => 'LPCWSTR');
+
+ $ffi->attach(takes_unicode_string => ['LPCWSTR'] => 'void');
+
+ my $str = "I ❤️ Platypus";
+ takes_unicode_string($str);
+
+=head1 DESCRIPTION
+
+This module provides a L<FFI::Platypus> custom type for Windows Unicode strings.
+The created LPCWSTR will be NULL-terminated.
+
+This type can also be used as a return value. Returned LPCWSTR are expected
+to be NULL-terminated, and they won't be freed.
+
+This type is automatically loaded by C<< $ffi->lang('Win32'); >>.
+
+=cut
+
+use constant _incantation =>
+  $^O eq 'MSWin32' && do { require Config; $Config::Config{archname} =~ /MSWin32-x64/ }
+  ? 'Q'
+  : 'L!';
+
+my @stack;  # To keep buffer alive.
+
+sub perl_to_native
+{
+  if(defined $_[0])
+  {
+    my $buf = encode('UTF-16le', $_[0]."\0");
+    push @stack, \$buf;
+    return unpack(_incantation, pack 'P', $buf);
+  }
+  else
+  {
+    push @stack, undef;
+    return undef;
+  }
+}
+
+sub perl_to_native_post
+{
+  pop @stack;
+}
+
+sub native_to_perl
+{
+   return unless defined $_[0];
+   return decode('UTF-16le', unpack('P'.(strlenW($_[0])*2), pack(_incantation, $_[0])));
```

Comment:
I think the libc function `wcslen`  or the windows native `lstrlenW` could be used here instead of strlenW.  Both return the number of characters if the string is in the BMP, and returns half the number of bytes if not.

It would mean that this code wouldn't be usable on non-windows but I don't think you would typically want to use it anywhere else.

Reply:
See below: think I am leaning toward making your `strlenW` function private.

Reply:
Some here, compute the encoding instead of hard coding it.